compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/DecomposeLinalgExtPackUnPackToAIR.cpp

```diff
@@ -0,0 +1,505 @@
+          packOp,
+          "require the tiled outer dimensions of the result are all 1s");
+    }
+
+    // 1. Use rank-reduced memref.subview op to extract the tile and untiled
+    // outer dims.
+    Location loc = packOp.getLoc();
+    Value input = getPackOpInputOrPaddedSource(rewriter, packOp);
+    auto inputShape = packOp.getInputType().getShape();
+    DenseMap<int64_t, OpFoldResult> dimAndTileMapping =
+        packOp.getDimAndTileMapping();
+    Attribute zeroIdxAttr = rewriter.getIndexAttr(0);
+    Attribute oneIdxAttr = rewriter.getIndexAttr(1);
+    SmallVector<OpFoldResult> readOffsets(srcRank, zeroIdxAttr);
+    SmallVector<OpFoldResult> readStrides(srcRank, oneIdxAttr);
+    SmallVector<OpFoldResult> readSizes;
+    SmallVector<int64_t> readShape;
+    for (auto i : llvm::seq<unsigned>(0, srcRank)) {
+      if (dimAndTileMapping.count(i)) {
+        readShape.push_back(getConstantIntValue(dimAndTileMapping[i])
+                                .value_or(ShapedType::kDynamic));
+        readSizes.push_back(dimAndTileMapping[i]);
+        continue;
+      }
+      if (ShapedType::isDynamic(inputShape[i])) {
+        assert(false);
+        // TODO: Dynamic input shape
+      } else {
+        readSizes.push_back(rewriter.getIndexAttr(inputShape[i]));
+      }
+      readShape.push_back(inputShape[i]);
+    }
+
+    Type elemType = packOp.getInputType().getElementType();
+    Attribute memorySpace =
+        packOp.getInputType().cast<MemRefType>().getMemorySpace();
+    auto readType = MemRefType::get(
+        readShape, elemType,
+        packOp.getInputType().cast<MemRefType>().getLayout(), memorySpace);
+
+    Value tile = rewriter.create<memref::SubViewOp>(
+        loc, readType, input, readOffsets, readSizes, readStrides);
+
+    // 2. Transpose the tile to match the inner tile order.
+
+    SmallVector<int64_t> perm = getPackUnpackRankReducedPerm(
+        inputShape, innerDimsPos, packOp.getOuterDimsPerm());
+
+    SmallVector<int64_t> transpShape = readShape;
+    applyPermutationToVector<int64_t>(transpShape, perm);
+
+    Value transposed = rewriter.create<memref::TransposeOp>(
+        loc, tile,
+        AffineMapAttr::get(
+            AffineMap::getPermutationMap(perm, packOp->getContext())));
+
+    // 3. Insert the inner tile to the destination.
+    int64_t destRank = packOp.getOutputRank();
+    SmallVector<OpFoldResult> writeStrides(destRank, oneIdxAttr);
+    SmallVector<OpFoldResult> writeOffsets(destRank, zeroIdxAttr);
+    SmallVector<OpFoldResult> writeSizes =
+        memref::getMixedSizes(rewriter, loc, packOp.getOutput());
+
+    memorySpace = packOp.getOutputType().cast<MemRefType>().getMemorySpace();
+    auto writeType = MemRefType::get(destShape, elemType, nullptr, memorySpace);
+    auto output_subview = rewriter.create<memref::SubViewOp>(
+        loc, writeType, packOp.getOutput(), writeOffsets, writeSizes,
+        writeStrides);
+
+    SmallVector<Value, 2> mt;
```

Comment:
Thanks for the comment. Will change "mt" to "emptyVec" to be more descriptive. Those are basically the dependency list (empty because dep analysis is in a later pass), offsets, sizes and strides fields for both src and dst. They are left empty for now and they will be populated in the AIR conversion pass in https://github.com/Xilinx/mlir-air/pull/395. 

Fundamentally this `air.dma` op being created here is simply a `memref.copy` with the more forgiving memref type checks. By creating `air.dma` instead of `memref.copy` I could skip over some `memref.expand_shape` and `memref.collapse_shape` (or rank-reducing `memref.subview`) here.